From b6df04ac0f5c50ea9bc41948c8dd5387be44c4c1 Mon Sep 17 00:00:00 2001
From: Elvis Lee <kwangwoong.lee@lge.com>
Date: Wed, 21 Aug 2019 20:44:57 +0900
Subject: [PATCH] Allow to have activeFocus for each window

This makes multiple windows focused if there is a single window per
screen assuming that windows across screens don't compete the keyboard
focus.

Change-Id: Id6f6fd0c95747b03d56c5e535f1313c27d67ab24
Upstream-Status: Inappropriate [webOS specific]
---
 src/quick/items/qquickwindow.cpp | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/src/quick/items/qquickwindow.cpp b/src/quick/items/qquickwindow.cpp
index d545d3b8e..f77f90480 100644
--- a/src/quick/items/qquickwindow.cpp
+++ b/src/quick/items/qquickwindow.cpp
@@ -939,6 +939,20 @@ static inline bool windowHasFocus(QQuickWindow *win)
     return win == focusWindow || QQuickRenderControl::renderWindowFor(win) == focusWindow;
 }
 
+static inline bool singleWindowOnScreen(QQuickWindow *win)
+{
+    const QWindowList windowList = QGuiApplication::allWindows();
+    for (int i = 0; i < windowList.count(); i++) {
+        QWindow *ii = windowList.at(i);
+        if (ii == win)
+            continue;
+        if (ii->screen() == win->screen())
+            return false;
+    }
+
+    return true;
+}
+
 /*!
 Set the focus inside \a scope to be \a item.
 If the scope contains the active focus item, it will be changed to \a item.
@@ -1013,7 +1027,8 @@ void QQuickWindowPrivate::setFocusInScope(QQuickItem *scope, QQuickItem *item, Q
     }
 
     if (!(options & DontChangeFocusProperty)) {
-        if (item != contentItem || windowHasFocus(q)) {
+        // Allow focused if there is only one window in the screen where it belongs.
+        if (item != contentItem || windowHasFocus(q) || singleWindowOnScreen(q)) {
             itemPrivate->focus = true;
             changed << item;
         }
